src/runtime/metadata.cpp

```diff
@@ -1,10 +1,56 @@
 #include "HalideRuntime.h"
+#include "scoped_mutex_lock.h"
 
-// This is a trick used to ensure that halide_filter_metadata_t (and related types)
-// are included in the runtime bitcode; since no (other) runtime function references them
-// they are ordinarily stripped, making Codegen_LLVM fail when we attempt to access
-// the types.
+extern "C" {
 
-#define INLINE inline __attribute__((weak)) __attribute__((used)) __attribute__((always_inline)) __attribute__((nothrow)) __attribute__((pure))
+struct _halide_runtime_internal_registered_filter_t {
+    struct _halide_runtime_internal_registered_filter_t *next;
+    const char *name;
+    const halide_filter_metadata_t* metadata;
+    int (*argv_func)(void **args);
+};
 
-INLINE void __force_include_halide_filter_metadata_t_types(halide_filter_metadata_t*) {}
+};
+
+namespace Halide { namespace Runtime { namespace Internal {
+
+struct list_head_t {
+    halide_mutex mutex;
+    _halide_runtime_internal_registered_filter_t *next;
+
+    list_head_t() : mutex(), next(NULL) {}
+};
+
+static list_head_t* get_list_head() {
+    static list_head_t head;
+    return &head;
+}
+
+} } }
+
+extern "C" {
+
+// This is looked up by name in Codegen_LLVM, which is easier to do
+// for functions with plain C linkage.
+WEAK void _halide_runtime_internal_register_metadata(_halide_runtime_internal_registered_filter_t *info) {
```

Comment:
There's some logic in LLVM_Runtime_Linker.cpp that avoids stripping halide_* functions from the module that won't trigger on this function due to the leading underscore. Will that be a problem?